### PR TITLE
Add created and updated dates for user, rooms, rooms_users

### LINKED ISSUE #26: Add dates
- [ ] Date created on Users
- [ ] Date updated on Users
- [ ] Date created Rooms
- [ ] Date updated Rooms

(write a migration)

### PR DESCRIPTION
**Pull request by frcroth**
Closes #26. Dates are automatically set to the current date on creation, updated date is automatically set on update statements.

Unfortunately I can not run the tests locally with the new 2 database setup.

### CODE REVIEW

**Comment by sirkrypt0** [comment]
> Closes #26. Dates are automatically set to the current date on creation, updated date is automatically set on update statements.
> 
> 
> 
> Unfortunately I can not run the tests locally with the new 2 database setup.

Why can't you run the tests locally?

**Comment by sirkrypt0** [comment]
> > Closes #26. Dates are automatically set to the current date on creation, updated date is automatically set on update statements.
> > 
> > 
> > 
> > Unfortunately I can not run the tests locally with the new 2 database setup.
> 
> Why can't you run the tests locally?

Does the procedure described [here](https://github.com/eisen-oxid/thermit/pull/41) not work?

File: server/migrations/2021-04-03-161950_add_dates/up.sql [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    FOR EACH ROW
+    EXECUTE PROCEDURE last_upd_trig();
+
+CREATE TRIGGER room_update_date_trigger
+    BEFORE UPDATE ON rooms
+    FOR EACH ROW
+    EXECUTE PROCEDURE last_upd_trig();
+
+CREATE TRIGGER rooms_updates_update_date_trigger
+    BEFORE UPDATE ON rooms
```
Comment (sirkrypt0) [inline server/migrations/2021-04-03-161950_add_dates/up.sql]: This should probably be
```suggestion
    BEFORE UPDATE ON rooms_users
```

File: server/migrations/2021-04-03-161950_add_dates/up.sql [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    BEFORE UPDATE ON users
+    FOR EACH ROW
+    EXECUTE PROCEDURE last_upd_trig();
+
+CREATE TRIGGER room_update_date_trigger
+    BEFORE UPDATE ON rooms
+    FOR EACH ROW
+    EXECUTE PROCEDURE last_upd_trig();
+
+CREATE TRIGGER rooms_updates_update_date_trigger
```
Comment (sirkrypt0) [inline server/migrations/2021-04-03-161950_add_dates/up.sql]:
```suggestion
CREATE TRIGGER rooms_users_update_date_trigger
```

**Review by sirkrypt0 — CHANGES_REQUESTED** [review]
Nice feature. I'd love to see some tests one that, like date gets set on creation and is updated on modification.

**Comment by frcroth** [comment]
I added a test that should not fail (the behavior works in production) but fails.

File: server/key.pem [hunk 3]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (sirkrypt0) [inline server/key.pem]: Is this intended here?

Comment (frcroth) [reply server/key.pem]: Oh committed by accident, but it was just an example anyway so nothing is lost

**Comment by frcroth** [comment]
Well if you @sirkrypt0  have any idea why the test fails, I'm happy to change this. It seems the trigger is not triggered in the test database for some reason. Tried debugging this but to no avail.

**Comment by sirkrypt0** [comment]
> Well if you @sirkrypt0 have any idea why the test fails, I'm happy to change this. It seems the trigger is not triggered in the test database for some reason. Tried debugging this but to no avail.

The database connection in the tests does never commit the changes to the database directly. This might be the issue here, as the database would only fill the columns once the data is committed.

File: server/migrations/2021-04-03-161950_add_dates/down.sql [hunk 4]
```diff
@@ -0,0 +1,9 @@
+ALTER TABLE users
+    DROP created,
+    DROP updated;
+
+ALTER TABLE rooms
+    DROP created,
+    DROP updated;
+
```
Comment (sirkrypt0) [inline server/migrations/2021-04-03-161950_add_dates/down.sql:8]:
```suggestion

ALTER TABLE rooms_users
    DROP COLUMN created,
    DROP COLUMN updated;

```

File: server/migrations/2021-04-03-161950_add_dates/down.sql [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+DROP FUNCTION last_upd_trig CASCADE; -- Also removes triggers
```
Comment (sirkrypt0) [inline server/migrations/2021-04-03-161950_add_dates/down.sql]:
```suggestion
DROP FUNCTION last_upd_trig CASCADE; -- Also removes triggers

```

**Review by sirkrypt0 — CHANGES_REQUESTED** [review]

**Comment by sirkrypt0** [comment]
> > Well if you @sirkrypt0 have any idea why the test fails, I'm happy to change this. It seems the trigger is not triggered in the test database for some reason. Tried debugging this but to no avail.
> 
> The database connection in the tests does never commit the changes to the database directly. This might be the issue here, as the database would only fill the columns once the data is committed.

I did some research and think this is actually not the problem here. The triggers should run once the event, which is an update in this case, fires, regardless if there is a transaction or not.

**Comment by sirkrypt0** [comment]
The trigger is executed correctly. This can be tested by changing the `last_upd_trig()` to also change the username like `NEW.username := 'updated';`. Running the test afterwards correctly changes the username. The problem is, that `current_timestamp` returns the current date and time with time zone, which is the time when the transaction starts. All database operations in a single test run in a single transaction, thus the `current_timestamp` is the same for all operations on a test connection.

File: server/migrations/2021-04-03-161950_add_dates/up.sql [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    ADD updated timestamp NOT NULL default current_timestamp;
+
+ALTER TABLE rooms_users
+    ADD created timestamp NOT NULL default current_timestamp,
+    ADD updated timestamp NOT NULL default current_timestamp;
+
+CREATE FUNCTION last_upd_trig() RETURNS trigger
+    LANGUAGE plpgsql AS
+$$BEGIN
+    NEW.updated := current_timestamp;
```
Comment (sirkrypt0) [inline server/migrations/2021-04-03-161950_add_dates/up.sql]: This should fix the issue with the database, as `current_timestamp` does not change during a transaction, whereas `clock_timestamp` does.

```suggestion
    NEW.updated := CLOCK_TIMESTAMP();
```

**Review by sirkrypt0 — CHANGES_REQUESTED** [review]

**Pull request merged**